src/artifact.rs

```diff
@@ -447,10 +447,36 @@ impl Artifact {
 
     }
 
+    /// Get set of non-import declarations that have not been defined. This must be an empty set in
+    /// order to `emit` the artifact.
+    pub fn undefined_symbols(&self) -> Vec<String> {
+        let mut decls = self.declarations.clone();
+        for def in self.definitions.iter() {
+            decls.remove(&def.name);
+        }
```

Comment:
This is algorithmically better :-), but I'm curious if there's a reason for preferring this over adding `defined` flags to `Decl`? 

A `defined` flag in `Decl` could also be used to diagnose a symbol being defined multiple times too, which seems like it would be nice.